Use custom data to save any browser integration related data

varjolintu opened the pull request:
## Description
Browser connection keys and Allow/Deny rules are stored to custom data instead of attributes.

## Motivation and context
Fixes https://github.com/keepassxreboot/keepassxc-browser/issues/11.

## How has this been tested?
Tested manually with the latest KeePassXC-Browser.

## Types of changes
- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

droidmonkey commented:
This needs to be rebased and cleaned up after the Custom Data PR is merged.

phoerious commented:
I wonder if there should be some transitional code for those who have used the old implementation. Maybe even check if we are using KDBX 4 already and then convert all string attributes to custom data, otherwise, keep using string attributes.

varjolintu commented:
This PR still needs a way to view/delete the stored keys. So a little more work is needed.

varjolintu commented:
Added the key view/removal to Browser Integration settings window and rebased.

varjolintu commented:
This feature will be added https://github.com/keepassxreboot/keepassxc/issues/1591 to the PR.

droidmonkey commented:
This also needs to convert existing attributes to custom data, or will you be supporting both?

varjolintu commented:
@droidmonkey I'll do a feature that converts both KeePassHTTP and KeePassXC-Browser attributes to custom data. I will not support both of them.

varjolintu commented:
Button added to Browser Integration settings page. Rebased.

droidmonkey commented:
Any reason for closing this?

phoerious commented:
Umm... I just deleted the release/2.3.1 branch and GitHub auto-closed this PR.

varjolintu commented:
So, merge against develop or 2.3.2?

droidmonkey commented:
Develop please

varjolintu commented:
Reminder: needs one modification after https://github.com/keepassxreboot/keepassxc/pull/1810 is merged.

varjolintu commented:
Updated and rebased.

giosh94mhz commented:
I've tested on my machine (Gentoo) against latest version and patched with this PR.

Compilation is ok, and the migration to property attributes executed successfully in a couple of DB with ~100-200 entries.

varjolintu commented:
Added one more thing: the shared key list is emptied after disconnecting the browsers. Previously the list was unchanged until visiting the settings page again.

varjolintu commented:
@droidmonkey I agree. This should be asked automatically (but still leaving the manual button just in case). If the KeePassXC-Browser Settings attribute is found, the conversion popup dialog appears. I'll do the necessary fixes today.

varjolintu commented:
I noticed that the keys are not converted to custom data, but instead still using the old KeePassXC-Browser entry method. Maybe this needs to be changed too.

varjolintu commented:
Key conversion done. (Please review the dialog text).

Every time database is opened, BrowserService class checks if there are any old attributes or old key entry. If found, a popup dialog is shown.

varjolintu commented:
Key count added to the popup dialog. Checking for legacy settings has been moved to another function.

droidmonkey commented on src/browser/BrowserOptionDialog.cpp:
```diff
@@ -150,3 +174,48 @@ void BrowserOptionDialog::showProxyLocationFileDialog()
+    if (QMessageBox::Yes != MessageBox::question(this,
+            tr("Delete the selected key?"),
+            tr("Do you really want to delete the selected key?\n"
+               "This may cause the affected plugins to malfunction."),
```
What does it mean to malfunction? I think the correct verbiage here is "This may prevent connection to the browser plugin."

Note that we warn when you remove an individual key, but when you click on "Disconnect all browsers" it just does it without any confirmation. Same with "Forget all remembered permissions" which is a little bit confusing, this button should probably say "Forget all site-specific settings on entries".

varjolintu replied:
I agree. There should be confirmation for each of these actions.

droidmonkey commented on src/browser/BrowserService.cpp:
```diff
@@ -459,17 +429,16 @@ void BrowserService::removeSharedEncryptionKeys()
-                                 tr("KeePassXC: Settings not available!"),
+    Database* db = getDatabase();
+    if (!db) {
+        QMessageBox::information(0, tr("KeePassXC: Settings not available!"),
```
Technically if you reach here you do not have an active database. I think this warning can be dropped.

droidmonkey commented on src/browser/BrowserService.cpp:
```diff
@@ -546,6 +513,80 @@ void BrowserService::removeStoredPermissions()
+        return;
+    }
+
+    Database* db = m_dbTabWidget->currentDatabaseWidget()->database();
```
Why isn't "getDatabase()" used here?

Here is my problem with this settings dialog, application settings are meant to be global. So these buttons make no sense at all. Basically I have to have the correct database selected prior to hitting the settings button, which is an undiscoverable "feature".

Here are some options:
1) Move everything below "Sort matching credentials..." into the database settings
2) Add a drop-down selection list above the buttons that let's you select from the currently open databases.

To me, option 1 is the most acceptable because it makes sense. It is also where the KeeShare settings are going so that would be consistent.

varjolintu replied:
A separate Browser Integration page should be added to the Database settings. Then it would be nice and clean.

droidmonkey commented on src/browser/BrowserService.cpp:
```diff
@@ -790,6 +831,75 @@ Database* BrowserService::getDatabase()
+    auto dialogResult = QMessageBox::warning(nullptr,
+                                             tr("KeePassXC: Legacy browser integration settings detected"),
+                                             tr("Legacy browser integration settings have been detected.\n"
+                                                "Do you want to transfer the settings and stored keys to custom data?\n"
```
I would just say "Do you want to upgrade the settings to the latest standard?\nThis is necessary to maintain compatibility with the browser plugin."

varjolintu commented:
An update:
- Moved database specific settings from Browser Intergration settings to Database settings
- Added confirmation popup dialog to the buttons before removing or converting any data
- Cleaned the code from Browser Integration side

varjolintu commented:
A final update:
- Added a warning message (and disabled the buttons etc.) to Database settings browser page when Browser Integration is disabled.

varjolintu commented:
One more update:
- Only keys with a prefix are shown in the Database settings custom data list
- Key prefix is stripped from the key string
